Modules/_hashopenssl.c

```diff
@@ -47,7 +47,9 @@
 #define PY_OPENSSL_HAS_SCRYPT 1
 #define PY_OPENSSL_HAS_SHA3 1
 #define PY_OPENSSL_HAS_SHAKE 1
+#if defined(NID_blake2s256) || defined(NID_blake2b512)
 #define PY_OPENSSL_HAS_BLAKE2 1
+#endif
```

Comment:
It looks like `PY_OPENSSL_HAS_BLAKE2` [isn't referenced anywhere](https://github.com/search?q=repo%3Apython%2Fcpython+PY_OPENSSL_HAS_BLAKE2&type=code), so I'm guessing as to what the "correct" behavior here is. Intuitively, I take this symbol to mean "linked libcrypto supports _some_ BLAKE2 functionality". Please correct me if this is wrong.

Reply:
If it builds against both libraries, this will work for now.  What we did in our BoringSSL patch set is blindly disable all of the unsupported hash functions: sha512_224 sha512_256, sha3, shake, and blake2 - per https://github.com/python/cpython/pull/116399/files#diff-b4fd2e5e447a6f18fb2508c350524c5cb1feed8ec92d70822eac54431ef71baf

I like this compile time feature detection idea better.

Reply:
In the update I pushed to the PR I changed the || to && as it is simpler to assume that if a library supports one of these categories it will support them all.  it'd probably cause a build failure if something only supported a fraction of a category - that can be dealt with if it ever arises (doubtful).

Reply:
eh.. || makes sense for blake2 and shake.  reverting that bit of my change. :)